core/src/test/java/org/apache/calcite/test/RelBuilderTest.java

```diff
@@ -2577,6 +2577,22 @@ private RelNode buildRelWithDuplicateAggregates(
     assertThat(ex.getMessage(), allOf(containsString("Expression"), containsString("not found")));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4409">[CALCITE-4409]
+   * Improve exception when RelBuilder tries to create a field on a non-struct expression</a>. */
+  @Test void testFieldOnNonStructExpression() {
+    final RelBuilder builder = RelBuilder.create(config().build());
+    IllegalStateException ex = assertThrows(IllegalStateException.class, () -> {
+      builder.scan("EMP")
+          .project(
+              builder.field(builder.field("EMPNO"), "abc"))
+          .build();
+    }, "Field should fail since we are trying access a field on expression with non-struct type");
+    assertThat(ex.getMessage(),
+        allOf(containsString("Trying to access field"),
+            containsString("in a type with no fields")));
```

Comment:
What is the purpose to split into two `containsString` here ?

Reply:
Simply because the actual fieldName value goes in between, and it is not really relevant in order to verify the exception message. I just followed the same pattern as the test just above.

Reply:
The field name does not change over time, so it would be better to use regular `is("....")` matcher to get better feedback in IDE and CI

Reply:
Just to confirm, the verification with `is(...)` would look like:
`assertThat(ex.getMessage(), is("Trying to access field abc in a type with no fields: SMALLINT"));`
Is that what you both would prefer?

Reply:
Exactly.

Reply:
Ok, changed.